docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+= Test Guidelines
+:toc:
+:toc-placement: preamble
+
+:sectnums:
+:experimental:
+
+This documents contains various information on testing the codebase. Over time it might
+grow to include sections on other testing strategies than unit or integration tests. 
+
+[[how-to-write-unit-tests]]
+== How to write unit tests?
+
+Instead of writing down general rules for how to write tests, this section presents a
+step-by-step approach on how to write tests for a very simply library file called `lib.go`
+that lives inside the `foo/bar` package.
+The library contains three one-line functions which means that we have to test three functions
+in order to gain 100% test coverage. For simplicity, the functions under test don't have
+any dependency on an external resource like a database connection or something else.
+Nevertheless we will present mechanisms to handle this as well.
+
+[[lib-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib.go
+----
+package bar // <1>
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Say returns the input string as is
+func Say(message string) string { // <2>
+	return fmt.Sprintf("%s", message)
+}
+
+// Shout returns the input string in upper case
+// letters with three exlamation marks.
+func Shout(message string) string { // <3>
+	return fmt.Sprintf("%s!!!", toUpper(message))
+}
+
+// toUpper returns the input string in upper case
+func toUpper(message string) string { // <4>
+	return strings.ToUpper(message)
+}
+----
+<1> The package name is `bar`. 
+<2> An exported function called `Say` that is unknown packages that import `bar`.
+<3> Another exported function called `Shout` that is unknown packages that import `bar`.
+<4> A function called `toUpper` that is unexported and unknown packages that import `bar`.
+
+=== Blackbox vs. Whitebox tests
+
+In a blackbox test, you test a package as if you **only have access to the
```

Comment:
It is about coverage and also about usage help I'd say. In principle somebody who would like to know how to use a package should take a look at the blackbox tests because this is the interface that one should be using. So separating the whitebox tests helps to bring some clarity for the reader.